### PR TITLE
Timer/hardclock fixes

### PR DESCRIPTION
**Pull request by gufe44**
- Use the RDTICK instruction instead of RDPR on sparc64 platforms: Apparently OpenBSD _does_ allow reads from the TICK register. The RDPR instruction is only available to privileged software however so this could not work on any operating system (nonprivileged software must use the RDTICK instruction instead).
- Add support for hardened systems with restricted access to the CPU cycle counter by reading system information when mbedtls_timing_hardclock is first called and using software clock if access is found to be restricted. This will still crash if access to the register is disabled afterwards. Currently supported: [NetBSD/i386 and amd64](https://github.com/NetBSD/src/commit/084a8872ba30ffdfbb690365326871adf409c4c9)
- Use monotonic time instead of wall time throughout the timer/hardclock module

### CODE REVIEW

**Comment by niacat** [comment]
I don't understand why `CLOCK_MONOTONIC` isn't used directly on NetBSD? TSC is unreliable on a bunch of systems, it's much better to let the kernel pick the most reasonable option from the range of sources it supports (or indeed the user with `sysctl kern.timecounter.`)

**Comment by bensze01** [comment]
The CI is complaining of the following errors:

```
timing.c:66:21: error: field ‘start’ has incomplete type
     struct timespec start;
                     ^
timing.c: In function ‘mbedtls_timing_get_timer’:
timing.c:353:9: error: implicit declaration of function ‘clock_gettime’ [-Werror=implicit-function-declaration]
         clock_gettime( CLOCK_MONOTONIC, &t->start );
         ^
timing.c:353:24: error: ‘CLOCK_MONOTONIC’ undeclared (first use in this function)
         clock_gettime( CLOCK_MONOTONIC, &t->start );
                        ^
timing.c:353:24: note: each undeclared identifier is reported only once for each function it appears in
timing.c:363:25: error: storage size of ‘now’ isn’t known
         struct timespec now;
                         ^
timing.c:363:25: error: unused variable ‘now’ [-Werror=unused-variable]
cc1: all warnings being treated as errors
Makefile:263: recipe for target 'timing.o' failed
make[1]: *** [timing.o] Error 1
make[1]: *** Waiting for unfinished jobs....
Makefile:17: recipe for target 'lib' failed
make: *** [lib] Error 2
^^^^test_gcc_opt: build/test: gcc -Os, full config: command make CC=gcc CFLAGS=-Os -std=c99 -pedantic -Wall -Wextra -Werror -> 2^^^^
(skipped because the build failed)
```

Full log: https://gist.github.com/bensze01/da8b24a393eebfd0066722cd30eb84e8

**Review by bensze01 — CHANGES_REQUESTED** [review]

**Comment by gufe44** [comment]
Thank you for pointing out the problems with _POSIX_TIMERS and _POSIX_MONOTONIC_CLOCK. I have taken a look on Debian, Solaris, NetBSD, OpenBSD and FreeBSD. Monotonic time is available on all of them but Debian sets _POSIX_MONOTONIC_CLOCK to 0 and OpenBSD (like macOS) sets _POSIX_TIMERS to -1. I concluded to implement runtime checks. The CI error was due to -std=c99 without defining _POSIX_C_SOURCE.

File: library/timing.c [hunk 1]
```diff
@@ -19,6 +19,18 @@
  *  This file is part of mbed TLS (https://tls.mbed.org)
  */
 
+/*
+ * for clock_gettime() from <time.h>
+ */
+#define _POSIX_C_SOURCE 200112L
```
Comment (bensze01) [inline library/timing.c:25]: [FreeBSD doesn't define `gettimeofday()` if only `_POSIX_C_SOURCE `is defined, without `_XOPEN_SOURCE`.](https://github.com/freebsd/freebsd/blob/release/11.1.0/sys/sys/time.h#L494)

See relevant failed build log [here](https://gist.github.com/bensze01/870cb75042140eb235962beb590dd6cf) (please ignore the warnings about `static_assert`).

`_XOPEN_SOURCE >= 600` also [implies `_POSIX_C_SOURCE 200112L`](https://man7.org/linux/man-pages/man7/feature_test_macros.7.html).
```suggestion
/*
 * for clock_gettime() from <time.h>
 * and for gettimeofday() from <sys/time.h> on FreeBSD
 */
#define _XOPEN_SOURCE 600
```

**Review by bensze01 — CHANGES_REQUESTED** [review]

**Comment by gufe44** [comment]
@niacat Could you clarify what your suggested change is?

**Comment by gufe44** [comment]
Unfortunately, for personal reasons I won't be able to continue to work on the PR. You can close the PR now and resurrect the branch at any time.

**Pull request closed without merging**